File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
     ListNode* reverseList(ListNode* head) {

        ListNode *a = head;

        if(head == nullptr){
            return head;
        }


        ListNode *b = head-> next;
        head->next = nullptr;
        ListNode *temp = nullptr;
        while(b != nullptr){
            temp = b->next;
            b->next = a;
            a = b;
            b = temp;
        }

        return a;

    }

};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head) -> None:
        temp = head
        n = 0
        while(temp != None):
            temp = temp.next
            n += 1

        mid = n//2
        mid_head = head
        for _ in range(mid):
            mid_head = mid_head.next


        a = mid_head
        b = a.next
        mid_head.next = None
        while(b != None):
            temp = b.next
            b.next = a
            a = b
            b = temp

        reverse_mid_head = a
        ans = head

        for _ in range(mid):
            temp1 = head.next
            temp2 = reverse_mid_head.next


            if(temp1 == reverse_mid_head):
                break

            head.next = reverse_mid_head
            reverse_mid_head.next = temp1
            

            head = temp1
            reverse_mid_head = temp2

        head = ans



        